### DISCUSSION TITLE
VSCode commnad to apply unsafe fix

### DISCUSSION THREAD

**Discussion by mattuylee** [biomejs/biome-vscode#421] [Ideas]
now it can automaticly apply safe fix when file save, but not unsafe fix. I'd like to bind a shortcut key to explicitly execute command like `biome check --write --unsafe ${file}`, but no vscode command found.
Is it posible to provide a command "biome.check" with its cli params? Or just something like "biome.fixUnsafe"?

I tried to run this shell script with the extension [Command Runner](https://marketplace.visualstudio.com/items?itemName=edonet.vscode-command-runner). It works but unacceptably slow, because it opens a terminal window firstly.

**Comment by boar-is** [comment]
+1. The plugin for WebStorm have this feature. So it's not consistent.

**Comment by ematipico** [comment]
That's on purpose and it won't change anytime soon unless there's a very strong reason to do so.

Unsafe fixes are unsafe for a reason, they can break your code, hence they require to be opted in, with intent.

You can still make unsafe fixes safe using the configuration: https://biomejs.dev/linter/#configure-the-rule-fix

**Reply by mattuylee** [reply]
Thanks for your reply~

I understand that unsafe fix should only be applied when users intentionally opt in.

Unsafe fixes are unsafe for a reason, so it's not sensible to make unsafe fixes safe. Under this circumstance, if someone still wants to apply unsafe fixes for a specific case, with intent, he can only run script manually, but cannot do it with a VSCode keyboard shortcut.

A new VSCode command like "biome.fixUnsafe" will allow users binding it to a keyboard shortcut. In the meantime, users who use this command must be with intent, as it won't triggered by default.

So I think this feature is useful. English is not my mother language, hope my expression is clear.

**Reply by ematipico** [reply]
Any LSP editor (VSCode too) already provides a keyboard shortcut (I think CMD + D) that shows all fixes (commonly called code actions) available inside a diagnostic range.  

Place the cursor inside the range of the diagnostic and hit the shortcut.

The editor will show a pop up where you can select the actions you want. Among those actions, there's also the unsafe fix.

**Reply by OliverJAsh** [reply]
I think it could still be useful to have something equivalent to `source.fixAll.biome`, i.e. a code action that doesn't require making a selection and instead runs on the whole source file.

**Comment by boar-is** [comment]
Thank you for the helpful link! At least I could opt in to something like converting `A[]` to `Array<A>`. I like to write the shorter version but prefer the longer.
